# PHPSTAN Result
```::error file=public/index.php,line=32,col=0::Function executeRoute() has parameter $action with no type specified.
::error file=public/index.php,line=32,col=0::Function executeRoute() has parameter $controller with no type specified.
::error file=public/index.php,line=32,col=0::Function executeRoute() has parameter $requestedMethod with no type specified.
::error file=public/index.php,line=32,col=0::Function executeRoute() has parameter $routes with no type specified.
::error file=public/index.php,line=59,col=0::Function isValidRoute() has parameter $action with no type specified.
::error file=public/index.php,line=59,col=0::Function isValidRoute() has parameter $controller with no type specified.
::error file=public/index.php,line=59,col=0::Function isValidRoute() has parameter $requestedMethod with no type specified.
::error file=public/index.php,line=59,col=0::Function isValidRoute() has parameter $routes with no type specified.
::error file=src/config/database.php,line=4,col=0::Function src\config\getConnection() has no return type specified.
::error file=src/controller/BaseController.php,line=10,col=0::Method src\controller\BaseController::redirectIfNotLoggedIn() has no return type specified.
::error file=src/controller/DashboardController.php,line=11,col=0::Method src\controller\DashboardController::getindex() has no return type specified.
::error file=src/controller/DashboardController.php,line=11,col=0::Method src\controller\DashboardController::getindex() has parameter $formdata with no type specified.
::error file=src/controller/HomeController.php,line=11,col=0::Method src\controller\HomeController::getindex() has no return type specified.
::error file=src/controller/HomeController.php,line=11,col=0::Method src\controller\HomeController::getindex() has parameter $formdata with no type specified.
::error file=src/controller/LoginController.php,line=17,col=0::Method src\controller\loginController::getLogin() has parameter $formdata with no type specified.
::error file=src/controller/LoginController.php,line=22,col=0::Method src\controller\loginController::postLogin() has no return type specified.
::error file=src/controller/LogoutController.php,line=11,col=0::Method src\controller\LogoutController::postLogout() has no return type specified.
::error file=src/controller/LogoutController.php,line=11,col=0::Method src\controller\LogoutController::postLogout() has parameter $formdata with no type specified.
::error file=src/controller/RegisterController.php,line=13,col=0::Method src\controller\RegisterController::getform() has no return type specified.
::error file=src/controller/RegisterController.php,line=13,col=0::Method src\controller\RegisterController::getform() has parameter $formdata with no type specified.
::error file=src/controller/RegisterController.php,line=17,col=0::Method src\controller\RegisterController::postform() has no return type specified.
::error file=src/controller/RegisterController.php,line=17,col=0::Method src\controller\RegisterController::postform() has parameter $formdata with no type specified.
::error file=src/controller/ViewController.php,line=6,col=0::Method src\controller\ViewController::__construct() has parameter $viewName with no type specified.
::error file=src/controller/_Template.php,line=11,col=0::Method src\controller\Template::template() has no return type specified.
::error file=src/controller/_Template.php,line=11,col=0::Method src\controller\Template::template() has parameter $formdata with no type specified.
::error file=src/helper/autoloader.php,line=2,col=0::Function customAutoloader() has no return type specified.
::error file=src/helper/autoloader.php,line=2,col=0::Function customAutoloader() has parameter $class with no type specified.
::error file=src/models/BaseModel.php,line=8,col=0::Property src\models\BaseModel::$connection has no type specified.
::error file=src/models/User.php,line=11,col=0::Method src\models\User::login() has no return type specified.
::error file=src/models/User.php,line=11,col=0::Method src\models\User::login() has parameter $password with no type specified.
::error file=src/models/User.php,line=11,col=0::Method src\models\User::login() has parameter $username with no type specified.
::error file=src/models/User.php,line=35,col=0::Method src\models\User::register() has no return type specified.
::error file=src/models/User.php,line=35,col=0::Method src\models\User::register() has parameter $email with no type specified.
::error file=src/models/User.php,line=35,col=0::Method src\models\User::register() has parameter $password with no type specified.
::error file=src/models/User.php,line=35,col=0::Method src\models\User::register() has parameter $username with no type specified.
::error file=src/models/_Template.php,line=11,col=0::Method src\models\Template::templateMethod() has no return type specified.
::error file=src/views/loginIndex.view.php,line=13,col=0::Variable $message might not be defined.
::error file=src/views/partials/header.view.php,line=17,col=0::Variable $title might not be defined.
::error file=src/views/partials/header.view.php,line=21,col=0::Variable $page might not be defined.
::error file=src/views/partials/header.view.php,line=21,col=0::Variable $page might not be defined.
::error file=src/views/partials/header.view.php,line=21,col=0::Variable $page might not be defined.
::error file=src/views/partials/header.view.php,line=28,col=0::File ends with a trailing whitespace. This may cause problems when running the code in the web browser. Remove the closing ?> mark or remove the whitespace.
::error file=src/views/partials/navbar.view.php,line=4,col=0::Variable $site might not be defined.
::error file=src/views/partials/navbar.view.php,line=5,col=0::Variable $site might not be defined.
::error file=src/views/partials/navbar.view.php,line=6,col=0::Variable $site might not be defined.
::error file=src/views/register.view.php,line=9,col=0::Variable $message might not be defined.
 ------ ----------------------------------------------------------------------- 
  Line   public/index.php                                                       
 ------ ----------------------------------------------------------------------- 
  32     Function executeRoute() has parameter $action with no type specified.  
  32     Function executeRoute() has parameter $controller with no type         
         specified.                                                             
  32     Function executeRoute() has parameter $requestedMethod with no type    
         specified.                                                             
  32     Function executeRoute() has parameter $routes with no type specified.  
  59     Function isValidRoute() has parameter $action with no type specified.  
  59     Function isValidRoute() has parameter $controller with no type         
         specified.                                                             
  59     Function isValidRoute() has parameter $requestedMethod with no type    
         specified.                                                             
  59     Function isValidRoute() has parameter $routes with no type specified.  
 ------ ----------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------- 
  Line   src/config/database.php                                            
 ------ ------------------------------------------------------------------- 
  4      Function src\config\getConnection() has no return type specified.  
 ------ ------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------- 
  Line   src/controller/BaseController.php                                     
 ------ ---------------------------------------------------------------------- 
  10     Method src\controller\BaseController::redirectIfNotLoggedIn() has no  
         return type specified.                                                
 ------ ---------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------- 
  Line   src/controller/DashboardController.php                               
 ------ --------------------------------------------------------------------- 
  11     Method src\controller\DashboardController::getindex() has no return  
         type specified.                                                      
  11     Method src\controller\DashboardController::getindex() has parameter  
         $formdata with no type specified.                                    
 ------ --------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------- 
  Line   src/controller/HomeController.php                                    
 ------ --------------------------------------------------------------------- 
  11     Method src\controller\HomeController::getindex() has no return type  
         specified.                                                           
  11     Method src\controller\HomeController::getindex() has parameter       
         $formdata with no type specified.                                    
 ------ --------------------------------------------------------------------- 

 ------ ----------------------------------------------------------------------- 
  Line   src/controller/LoginController.php                                     
 ------ ----------------------------------------------------------------------- 
  17     Method src\controller\loginController::getLogin() has parameter        
         $formdata with no type specified.                                      
  22     Method src\controller\loginController::postLogin() has no return type  
         specified.                                                             
 ------ ----------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------- 
  Line   src/controller/LogoutController.php                                 
 ------ -------------------------------------------------------------------- 
  11     Method src\controller\LogoutController::postLogout() has no return  
         type specified.                                                     
  11     Method src\controller\LogoutController::postLogout() has parameter  
         $formdata with no type specified.                                   
 ------ -------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------- 
  Line   src/controller/RegisterController.php                               
 ------ -------------------------------------------------------------------- 
  13     Method src\controller\RegisterController::getform() has no return   
         type specified.                                                     
  13     Method src\controller\RegisterController::getform() has parameter   
         $formdata with no type specified.                                   
  17     Method src\controller\RegisterController::postform() has no return  
         type specified.                                                     
  17     Method src\controller\RegisterController::postform() has parameter  
         $formdata with no type specified.                                   
 ------ -------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------- 
  Line   src/controller/ViewController.php                                  
 ------ ------------------------------------------------------------------- 
  6      Method src\controller\ViewController::__construct() has parameter  
         $viewName with no type specified.                                  
 ------ ------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------- 
  Line   src/controller/_Template.php                                        
 ------ -------------------------------------------------------------------- 
  11     Method src\controller\Template::template() has no return type       
         specified.                                                          
  11     Method src\controller\Template::template() has parameter $formdata  
         with no type specified.                                             
 ------ -------------------------------------------------------------------- 

 ------ --------------------------------------------------------------- 
  Line   src/helper/autoloader.php                                      
 ------ --------------------------------------------------------------- 
  2      Function customAutoloader() has no return type specified.      
  2      Function customAutoloader() has parameter $class with no type  
         specified.                                                     
 ------ --------------------------------------------------------------- 

 ------ ------------------------------------------------------------------- 
  Line   src/models/BaseModel.php                                           
 ------ ------------------------------------------------------------------- 
  8      Property src\models\BaseModel::$connection has no type specified.  
 ------ ------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------- 
  Line   src/models/User.php                                                   
 ------ ---------------------------------------------------------------------- 
  11     Method src\models\User::login() has no return type specified.         
  11     Method src\models\User::login() has parameter $password with no type  
         specified.                                                            
  11     Method src\models\User::login() has parameter $username with no type  
         specified.                                                            
  35     Method src\models\User::register() has no return type specified.      
  35     Method src\models\User::register() has parameter $email with no type  
         specified.                                                            
  35     Method src\models\User::register() has parameter $password with no    
         type specified.                                                       
  35     Method src\models\User::register() has parameter $username with no    
         type specified.                                                       
 ------ ---------------------------------------------------------------------- 

 ------ ----------------------------------------------------------------- 
  Line   src/models/_Template.php                                         
 ------ ----------------------------------------------------------------- 
  11     Method src\models\Template::templateMethod() has no return type  
         specified.                                                       
 ------ ----------------------------------------------------------------- 

 ------ ----------------------------------------- 
  Line   src/views/loginIndex.view.php            
 ------ ----------------------------------------- 
  13     Variable $message might not be defined.  
 ------ ----------------------------------------- 

 ------ -------------------------------------------------------------------- 
  Line   src/views/partials/header.view.php                                  
 ------ -------------------------------------------------------------------- 
  17     Variable $title might not be defined.                               
  21     Variable $page might not be defined.                                
  21     Variable $page might not be defined.                                
  21     Variable $page might not be defined.                                
  28     File ends with a trailing whitespace. This may cause problems when  
         running the code in the web browser. Remove the closing ?> mark or  
         remove the whitespace.                                              
 ------ -------------------------------------------------------------------- 

 ------ -------------------------------------- 
  Line   src/views/partials/navbar.view.php    
 ------ -------------------------------------- 
  4      Variable $site might not be defined.  
  5      Variable $site might not be defined.  
  6      Variable $site might not be defined.  
 ------ -------------------------------------- 

 ------ ----------------------------------------- 
  Line   src/views/register.view.php              
 ------ ----------------------------------------- 
  9      Variable $message might not be defined.  
 ------ ----------------------------------------- 

 [ERROR] Found 46 errors                                                        
```

--- a/public/index.php
+++ b/public/index.php
@@ -1,4 +1,5 @@
 <?php
+
 use src\controller\ViewController;
 use src\helper;
 
@@ -13,27 +14,29 @@ $requestedUri = ($_SERVER['REQUEST_URI']);
 $requestedMethod = ($_SERVER['REQUEST_METHOD']);
 // split the entered uri into single fragments
 $splittedUri = explode('/', $requestedUri);
-$splittedUri[1]!="" ? $controller = $splittedUri[1] : $controller="home";
-if(count($splittedUri)>2){
+$splittedUri[1] != "" ? $controller = $splittedUri[1] : $controller = "home";
+if (count($splittedUri) > 2) {
     $action = $splittedUri[2];
-}else{
+} else {
     $action = '';
 }
 
 executeRoute($controller, $action, $routes, $requestedMethod);
 
 /**
- * @param $controller
- * @param $action
- * @param $routes
- * @param $requestedMethod
+ * redirect to the correct controller and action
+ *
+ * @param string $controller
+ * @param string $action
+ * @param mixed $routes
+ * @param mixed $requestedMethod
  * @return void
  */
-function executeRoute($controller, $action, $routes, $requestedMethod)
+function executeRoute(string $controller, string $action, mixed $routes, mixed $requestedMethod): void
 {
     // Trying to get the routes endpoint if the route is completely defined in the routesfile
     $routeIsValid = isValidRoute($controller, $action, $routes, $requestedMethod);
-    if ($routeIsValid){
+    if ($routeIsValid) {
         // fetch internal functionname
         $controllerFunction = strtolower($requestedMethod) . $routes[$controller][$requestedMethod][$action];
         // prepare Controller namespace for calling the Class
@@ -48,17 +51,17 @@ function executeRoute($controller, $action, $routes, $requestedMethod)
 }
 
 /**
- * @param $controller
- * @param $action
- * @param $routes
- * @param $requestedMethod
- * @return bool
- *
  * This function will test if there is a defined internal route based on the wanted url parameter
+ *
+ * @param string $controller
+ * @param string $action
+ * @param mixed $routes
+ * @param mixed $requestedMethod
+ * @return bool
  */
-function isValidRoute($controller, $action, $routes, $requestedMethod)
+function isValidRoute(string $controller, string $action, mixed $routes, mixed $requestedMethod): bool
 {
-// check if route is defined in routes.php
+    // check if route is defined in routes.php
     $controllerExists = in_array($controller, (array_keys($routes)));
     if (!$controllerExists) {
         return false; // stop here if controller is not correct

--- a/src/config/database.php
+++ b/src/config/database.php
@@ -1,7 +1,13 @@
 <?php
 namespace src\config;
 
-function getConnection(){
-$connection=mysqli_connect('localhost', 'shop', '1234', 'shop');
-return $connection;
+use mysqli;
+
+/**
+ * create connection to database
+ *
+ * @return mysqli|false
+ */
+function getConnection(): mysqli|false{
+    return mysqli_connect('localhost', 'shop', '1234', 'shop');
 }

--- a/src/controller/BaseController.php
+++ b/src/controller/BaseController.php
@@ -7,7 +7,8 @@ class BaseController
     {
     }
 
-    protected function redirectIfNotLoggedIn(){
+    protected function redirectIfNotLoggedIn(): void
+    {
         session_start();
         if(!isset($_SESSION['user'])){
             header('location: /login', true , 302);

--- a/src/controller/DashboardController.php
+++ b/src/controller/DashboardController.php
@@ -5,10 +5,11 @@ class DashboardController extends BaseController
 {
     public function __construct()
     {
+        parent::__construct();
         parent::redirectIfNotLoggedIn();
     }
 
-    public function getindex($formdata=null)
+    public function getindex($formdata=null): void
     {
         new ViewController('dashboard');
     }

--- a/src/controller/HomeController.php
+++ b/src/controller/HomeController.php
@@ -4,11 +4,11 @@ namespace src\controller;
 class HomeController extends BaseController
 {
     public function __construct()
-
     {
+        parent::__construct();
     }
 
-    public function getindex($formdata=null)
+    public function getindex(mixed $formdata=null): void
     {
         new ViewController('home');
     }

--- a/src/controller/LoginController.php
+++ b/src/controller/LoginController.php
@@ -6,6 +6,7 @@ class loginController extends BaseController
 {
     public function __construct()
     {
+        parent::__construct();
     }
 
     /**
@@ -14,12 +15,12 @@ class loginController extends BaseController
      * @param $formdata
      * @return void
      */
-    public function getLogin($formdata=null)
+    public function getLogin($formdata=null): void
     {
         new ViewController('loginIndex');
     }
 
-    public function postLogin()
+    public function postLogin(): void
     {
         $username = $_REQUEST['username'];
         $password = $_REQUEST['password'];

--- a/src/controller/LogoutController.php
+++ b/src/controller/LogoutController.php
@@ -4,11 +4,11 @@ namespace src\controller;
 class LogoutController extends BaseController
 {
     public function __construct()
-
     {
+        parent::__construct();
     }
 
-    public function postLogout($formdata=null)
+    public function postLogout(mixed $formdata=null): void
     {
         session_start();
         unset($_SESSION['user']);

--- a/src/controller/RegisterController.php
+++ b/src/controller/RegisterController.php
@@ -6,15 +6,15 @@ use src\models\User;
 class RegisterController extends BaseController
 {
     public function __construct()
-
     {
+        parent::__construct();
     }
 
-    public function getform($formdata=null)
+    public function getform($formdata=null): void
     {
         new ViewController('register');
     }
-    public function postform($formdata=null)
+    public function postform($formdata=null): void
     {
         $username=$_REQUEST['username'];
         $password=$_REQUEST['password'];

--- a/src/controller/ViewController.php
+++ b/src/controller/ViewController.php
@@ -5,6 +5,7 @@ class ViewController extends BaseController
 {
     public function __construct($viewName)
     {
+        parent::__construct();
         include_once("..\\src\\views\\{$viewName}.view.php");
     }
 }

--- a/src/controller/_Template.php
+++ b/src/controller/_Template.php
@@ -4,11 +4,11 @@ namespace src\controller;
 class Template extends BaseController
 {
     public function __construct()
-
     {
+        parent::__construct();
     }
 
-    public function template($formdata=null)
+    public function template(mixed $formdata=null): void
     {
     }
 

--- a/src/helper/autoloader.php
+++ b/src/helper/autoloader.php
@@ -1,5 +1,6 @@
 <?php
-function customAutoloader($class){
+function customAutoloader(mixed $class): void
+{
     $file = str_replace('\\', '/', $class) . '.php';
     $path=__DIR__.'/../../';
     include_once($path.$file);

--- a/src/models/BaseModel.php
+++ b/src/models/BaseModel.php
@@ -5,7 +5,7 @@ include_once("../src/config/database.php");
 
 class BaseModel
 {
-    protected $connection;
+    protected \mysqli|false $connection;
 
     protected function __construct()
     {

--- a/src/models/User.php
+++ b/src/models/User.php
@@ -1,4 +1,5 @@
 <?php
+
 namespace src\models;
 
 class User extends BaseModel
@@ -8,56 +9,70 @@ class User extends BaseModel
         parent::__construct();
     }
 
-    public function login($username,$password)
+    /**
+     * Login the user
+     *
+     * @param string $username
+     * @param string $password
+     * @return void
+     */
+    public function login(string $username, string $password): void
     {
         try {
-        $query = "select * from users where name = '{$username}' limit 1";
-        $sql=$this->connection->query($query);
-        $result=$sql->fetch_object();
-        // Check if there is a user with the send login mail or username
-        if(password_verify($password,$result->password)){
-            // if everything is ok perform login and set user as active user for the session
-            session_start();
-            $_SESSION['user'] = $result->id;
-            header("location : {$_SERVER['HTTP_ORIGIN']}/dashboard",true,302);
-        }
-        else{
-            throw new \Exception('login fehlgeschlagen');
-        }
-        }catch (\Exception $exception){
+            $query = "select * from users where name = '{$username}' limit 1";
+            $sql = $this->connection->query($query);
+            $result = $sql->fetch_object();
+            // Check if there is a user with the send login mail or username
+            if (password_verify($password, $result->password)) {
+                // if everything is ok perform login and set user as active user for the session
+                session_start();
+                $_SESSION['user'] = $result->id;
+                header("location : {$_SERVER['HTTP_ORIGIN']}/dashboard", true, 302);
+            } else {
+                throw new \Exception('login fehlgeschlagen');
+            }
+        } catch (\Exception $exception) {
             session_unset();
             session_start();
             $_SESSION['message'] = $exception->getMessage();
-            header("location : {$_SERVER['HTTP_ORIGIN']}/login",true,302);
+            header("location : {$_SERVER['HTTP_ORIGIN']}/login", true, 302);
         }
     }
 
-    public function register($username,$password,$email)
+    /**
+     * Register a new user
+     *
+     * @param string $username
+     * @param string $password
+     * @param string $email
+     * @return void
+     */
+    public function register(string $username, string $password, string $email): void
     {
-     // Check if email is already taken
+        // Check if email is already taken
         try {
             //limiting to have less stress on database
             $query = "select * from users where email = '{$email}' limit 1";
-            $sql=$this->connection->query($query);
-            $result=$sql->fetch_object();
-            $result=$sql->num_rows;
-            if($result==1){
+            $sql = $this->connection->query($query);
+            $result = $sql->fetch_object();
+            $result = $sql->num_rows;
+            if ($result == 1) {
                 session_start();
                 $_SESSION['message'] = "Email bereits vergeben";
-                header('location : /register' ,true, 302);
-            }else{
+                header('location : /register', true, 302);
+            } else {
                 $query = "Insert INTO users (name,password,email) values ('{$username}','{$password}','{$email}')";
-                $saved=$sql=$this->connection->query($query);
-                if(!$saved){
+                $saved = $sql = $this->connection->query($query);
+                if (!$saved) {
                     $_SESSION['message'] = "Hoppla, da ist etwas schiefgelaufen";
-                    header("location: /register" ,true,302);
+                    header("location: /register", true, 302);
                 }
-                $userId=$this->connection->insert_id; // get id after creation
+                $userId = $this->connection->insert_id; // get id after creation
                 $_SESSION['user'] = $userId; // login aver successful creation
-                header("location: /dashboard" ,true,302);
+                header("location: /dashboard", true, 302);
             }
 
-    }catch (\Exception $e ){
+        } catch (\Exception $e) {
             var_dump($e);
         }
     }

--- a/src/models/_Template.php
+++ b/src/models/_Template.php
@@ -8,7 +8,7 @@ class Template extends BaseModel
         parent::__construct();
     }
 
-    public function templateMethod()
+    public function templateMethod(): void
     {
 
     }

--- a/src/views/loginIndex.view.php
+++ b/src/views/loginIndex.view.php
@@ -1,4 +1,5 @@
 <?php
+global $message;
 $header = __DIR__ . "\\partials\\header.view.php";
 // Titel der Seite eintragen
 $title = "login";

--- a/src/views/partials/header.view.php
+++ b/src/views/partials/header.view.php
@@ -1,4 +1,5 @@
 <?php
+global $title, $page;
 if(!isset($_SESSION)) // avoid double opening sessions
 {
     session_start();
@@ -23,6 +24,5 @@ if ($page!="login" && $page!= 'register' && $page!="home"){ // avoid showing hea
         include_once('navbar.view.php');
     }
 }
-
 ?>
 

--- a/src/views/register.view.php
+++ b/src/views/register.view.php
@@ -1,4 +1,5 @@
 <?php
+global $message;
 $header=__DIR__."\\partials\\header.view.php";
 // Titel der Seite eintragen
 $title = "Registrieren";

